README.md
# netbeans-intellij-plugin
This is an Intellij plugin, which helps you integrate Netbeans project into Intellij

1. Add new module from Netbeans source: 
	- auto import dependency listed in nbproject/project.properties into Intellij Module
	- create Intellij Module Run Configuration
	- create Ant build jar for Intellij Module
2. When add/remove Intellij module dependency, auto update nbproject/project.properties



CHANGELOG.md
1.0-201706111300
- Auto-update nbproject/project.properties whenever new dependency is added to Intellij module


1.0-201705171500
- Auto-resolve libraries when import Netbeans project as a Intellij Module.


src/converter/ConverterFactory.java
/*
 * Copyright 2000-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package converter;

/**
 * Created by trangdp on 16/05/2017.
 */
public class ConverterFactory {

    public ProjectFileConverter getConverter(String type) {
        switch(type) {
            case "netbeans-to-intellij":
                return new NetbeansToIntellijConverter();
            case "intellij-to-netbeans":
                return new IntellijToNetbeansConverter();
            default:
                return new NetbeansToIntellijConverter();
        }
    }

}


src/converter/ProjectFileConverter.java
/*
 * Copyright 2000-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package converter;

import com.intellij.openapi.module.Module;

/**
 * Created by trangdp on 16/05/2017.
 */
public interface ProjectFileConverter {
  default void setModule(Module module) {}
  default void setProjectFileContent(String content) {}
  default void convert() {}
}


src/converter/NetbeansFileConverter.java
/*
 * Copyright 2000-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package converter;

import com.intellij.openapi.module.Module;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import org.jetbrains.annotations.NotNull;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import util.NotificationUtil;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpression;
import javax.xml.xpath.XPathFactory;
import java.io.*;
import java.util.Properties;
import java.util.Set;

/**
 * Created by trangdp on 16/05/2017.
 */
public class NetbeansFileConverter implements ProjectFileConverter {
    @NotNull
    private Module module;

    @NotNull
    private String projectFileContent;

    public void setModule(Module module) {
        this.module = module;
    }

    public void setProjectFileContent(String projectFileContent) {
        this.projectFileContent = projectFileContent;
    }

    public void convert() {
        NotificationUtil.notify("Converting " + module.getName() + " ...");

        byte[] contentWithLibraryRemoved = clearIntellijModuleLibrary();

        if (contentWithLibraryRemoved != null) {
            InputStream is = new ByteArrayInputStream(projectFileContent.getBytes());
            String moduleLibraries = "<root>";
            try {
                Properties properties = new Properties();
                properties.load(is);

                Set<String> propertyNames = properties.stringPropertyNames();
                for (String name : propertyNames) {
                    if (name.startsWith("javac.classpath")) {
                        String classPath = properties.getProperty(name);
                        String[] referenceClasses = classPath.split(":");
                        for (String reference : referenceClasses) {
                            reference = reference.replace("${", "").replace("}", "");
                            String pathToJar = properties.getProperty(reference);
                            if (pathToJar.startsWith("lib/") || pathToJar.startsWith("libs/")) {
                                pathToJar = "$MODULE_DIR$/" + pathToJar;
                                NotificationUtil.notify(pathToJar);
                            }

                            moduleLibraries += extractNetbeansLibrary(pathToJar);
                        }
                    }
                }

                moduleLibraries += "</root>";

                resolveIntellijLibrary(moduleLibraries, contentWithLibraryRemoved);

            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        NotificationUtil.notify("Converted " + module.getName() + " DONE");
    }

    private byte[] clearIntellijModuleLibrary() {
        if (module == null) {
            return null;
        }

        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder = null;
        FileOutputStream os = null;

        try {
            PsiFile moduleImlFile = getModuleImlFile();
            InputStream inputStream = moduleImlFile.getVirtualFile().getInputStream();

            dBuilder = dbFactory.newDocumentBuilder();
            Document doc = dBuilder.parse(inputStream);

            inputStream.close();

            doc.getDocumentElement().normalize();
            XPathFactory xpf = XPathFactory.newInstance();
            XPath xpath = xpf.newXPath();
            XPathExpression expression = xpath.compile("//module/component/orderEntry[@type=\"module-library\"] | //module/component/orderEntry[@level=\"project\"]");
            NodeList node = (NodeList) expression.evaluate(doc, XPathConstants.NODESET);

            for (int i = 0; i < node.getLength(); i++) {
                Node item = node.item(i);
                item.getParentNode().removeChild(item);
            }

            TransformerFactory tf = TransformerFactory.newInstance();
            Transformer t = tf.newTransformer();
            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();

            t.transform(new DOMSource(doc), new StreamResult(byteArrayOutputStream));
            t.reset();

            return byteArrayOutputStream.toByteArray();
        } catch (Exception e) {
            e.printStackTrace();
        }

        return null;
    }

    private String extractNetbeansLibrary(@NotNull String pathToJar) {
        String template = "" +
                "<orderEntry type=\"module-library\">\n" +
                "   <library>\n" +
                "       <CLASSES>\n" +
                "           <root url=\"jar://%s!/\" />\n" +
                "       </CLASSES>\n" +
                "       <JAVADOC />\n" +
                "       <SOURCES />\n" +
                "   </library>\n" +
                "</orderEntry>\n";
        String configuration = String.format(template, pathToJar);

        return configuration;
    }

    private void resolveIntellijLibrary(String input, byte[] contentWithLibraryRemoved) throws IOException {
        if (module == null) {
            return;
        }

        if(!isModuleImlFileExist()) {
            Messages.showInfoMessage(module.getModuleFilePath(), "Intellij Project file not found.");
            return;
        }

        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder = null;
        FileOutputStream os = null;

        try {

            InputStream inputStream = new ByteArrayInputStream(contentWithLibraryRemoved);
            dBuilder = dbFactory.newDocumentBuilder();
            Document doc = dBuilder.parse(inputStream);

            inputStream.close();

            doc.getDocumentElement().normalize();
            XPathFactory xpf = XPathFactory.newInstance();
            XPath xpath = xpf.newXPath();
            XPathExpression expression = xpath.compile("/module/component");
            Node node = (Node) expression.evaluate(doc, XPathConstants.NODE);

            if (node != null) {
                Node dependencyRoot = dBuilder
                        .parse(new ByteArrayInputStream(input.getBytes())).getDocumentElement();
                NodeList childNodes = dependencyRoot.getChildNodes();
                for (int i = 0; i < childNodes.getLength(); ++i) {
                    Node dependency = childNodes.item(i);
                    dependency = doc.importNode(dependency, true);
                    node.appendChild(dependency);
                }
            }

            os = new FileOutputStream(new File(module.getModuleFilePath()), false);
            TransformerFactory tf = TransformerFactory.newInstance();
            Transformer t = tf.newTransformer();
            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();

            t.transform(new DOMSource(doc), new StreamResult(byteArrayOutputStream));
            t.reset();

            os.write(byteArrayOutputStream.toByteArray());
            os.flush();

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (os != null) try {
                os.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private boolean isModuleImlFileExist() {
        String relPath = module.getModuleFilePath().substring(0, module.getModuleFilePath().lastIndexOf(File.separator));
        String path = relPath + File.separator + module.getName() + ".iml";
        VirtualFile moduleImlFile = module.getProject().getBaseDir().getFileSystem().findFileByPath(path);

        return (moduleImlFile != null);
    }


    private PsiFile getModuleImlFile() {
        String relPath = module.getModuleFilePath().substring(0, module.getModuleFilePath().lastIndexOf(File.separator));
        String path = relPath + File.separator + module.getName() + ".iml";
        VirtualFile moduleImlFile = module.getProject().getBaseDir().getFileSystem().findFileByPath(path);

        PsiFile moduleFile = PsiManager.getInstance(module.getProject()).findFile(moduleImlFile);
        return moduleFile;
    }
}


src/converter/IntellijModuleImlChecksum.java
package converter;

import org.apache.commons.codec.digest.DigestUtils;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Created by trangdp on 17/05/2017.
 */
public class IntellijModuleImlChecksum {
    private static IntellijModuleImlChecksum Instance = null;
    private ConcurrentHashMap<String, String> checksumCache = new ConcurrentHashMap<>();

    private IntellijModuleImlChecksum() {
    }

    public static IntellijModuleImlChecksum getInstance() {
        if (Instance == null) {
            Instance = new IntellijModuleImlChecksum();
        }

        return Instance;
    }

    public String getLastCheckSum(String moduleName) {
        return checksumCache.get(moduleName);
    }

    public void updateLastChecksum(String moduleName, String pathToFile) throws IOException {
        String md5 = checksum(pathToFile);
        this.checksumCache.put(moduleName, md5);
    }

    public String checksum(String pathToFile) throws IOException {
        InputStream is = Files.newInputStream(Paths.get(pathToFile));
        String md5 = DigestUtils.md5Hex(is);

        return md5;

    }

}


src/converter/IntellijToNetbeansConverter.java
package converter;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.OrderEntry;
import com.intellij.openapi.roots.OrderEnumerator;
import com.intellij.openapi.roots.impl.ModuleLibraryOrderEntryImpl;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.util.Processor;
import org.jetbrains.annotations.NotNull;
import util.NotificationUtil;

import java.io.*;
import java.util.*;

/**
 * Created by trangdp on 17/05/2017.
 * This class is to convert intellij dependecies to netbeans
 */
public class IntellijToNetbeansConverter implements ProjectFileConverter {
    private final Logger logger = Logger.getInstance("com.trangdp.IntellijToNetbeansConverter");

    @NotNull
    private Module module;

    @NotNull
    private String projectFileContent;

    @Override
    public void setModule(@NotNull Module module) {
        this.module = module;
    }

    @Override
    public void setProjectFileContent(@NotNull String projectFileContent) {
        this.projectFileContent = projectFileContent;
    }

    @Override
    public void convert() {
        NotificationUtil.notify("Intellij to Netbeans converting...");

        //TODO:clear netbeans classpath, read orderEntry module-library intellij, re-build netbeans classpath
        Properties properties = clearNetbeansClasspath();

        if(properties != null) {
            List<String> references = extractIntellijReferences();
            Properties transformedNetbeansClasspath = buildNetbeansClasspath(properties, references);

            try {
                flushNetbeansClasspath(transformedNetbeansClasspath);
            } catch (IOException e) {
                logger.error("Unable to override netbeans project properties");
            }
        }

        NotificationUtil.notify("Intellij to Netbeans converted DONE");
    }

    private void flushNetbeansClasspath(Properties transformedNetbeansClasspath) throws IOException {
        Project project = module.getProject();
        String relPath = module.getModuleFilePath().substring(0, module.getModuleFilePath().lastIndexOf(File.separator));
        VirtualFile moduleRoot = project.getBaseDir().getFileSystem().findFileByPath(relPath);
        if (moduleRoot.isDirectory()) {
            String netbeansPropertiesPath = String.format("%s/nbproject/project.properties", moduleRoot.getCanonicalPath());

            PrintWriter writer = new PrintWriter(netbeansPropertiesPath);
            try {
                writer.print("");
                Enumeration propertiesKeys = sortPropertiesKeys(transformedNetbeansClasspath);
                while(propertiesKeys.hasMoreElements()) {
                    Object propertiesKey = propertiesKeys.nextElement();
                    writer.println(String.format("%s=%s", propertiesKey, transformedNetbeansClasspath.get(propertiesKey)));
                }

//                for (String name: transformedNetbeansClasspath.stringPropertyNames()) {
//                    writer.println(String.format("%s=%s", name, transformedNetbeansClasspath.get(name)));
//                }
                writer.flush();
            } finally {
                if(writer != null) writer.close();
            }
        }
    }

    private Properties buildNetbeansClasspath(Properties properties, List<String> references) {
        String javacClassPath = "\\\r\n";
        for (int i = 0; i < references.size(); ++i) {
            String reference = references.get(i);
            String[] split = reference.split("/");
            String libraryName = split[split.length - 1];

            properties.put("file.reference." + libraryName, reference);
            if(i < references.size() - 1) {
                javacClassPath += String.format("\t${file.reference.%s}:\\\r\n", libraryName);
            } else {
                javacClassPath += String.format("\t${file.reference.%s}", libraryName);
            }
        }

        properties.put("javac.classpath", javacClassPath);

        return properties;
    }

    private Enumeration sortPropertiesKeys(Properties properties) {
        Enumeration<?> keysEnum = properties.propertyNames();

        Vector keyList = new Vector();
        while(keysEnum.hasMoreElements()){
            keyList.add(keysEnum.nextElement());
        }
        Collections.sort(keyList);

        return keyList.elements();
    }

    private List<String> extractIntellijReferences() {
        List<String> references = new ArrayList<String>();
        String moduleName = module.getName();

        OrderEnumerator.orderEntries(module).forEach(new Processor<OrderEntry>() {
            @Override
            public boolean process(OrderEntry orderEntry) {
                if(orderEntry instanceof ModuleLibraryOrderEntryImpl) {
                    System.out.println(orderEntry.getPresentableName());
                    String libraryReference = orderEntry.getPresentableName();

                    if(libraryReference.contains(moduleName)) {
                        String[] split = libraryReference.split(moduleName + "/");
                        references.add(split[1]);
                    } else {
                        references.add(libraryReference);
                    }
                }

                return true;
            }
        });

        return references;
    }

    private Properties clearNetbeansClasspath() {
        Project project = module.getProject();
        String relPath = module.getModuleFilePath().substring(0, module.getModuleFilePath().lastIndexOf(File.separator));
        VirtualFile moduleRoot = project.getBaseDir().getFileSystem().findFileByPath(relPath);
        if (moduleRoot.isDirectory()) {
            String netbeansPropertiesPath = String.format("%s/nbproject/project.properties", moduleRoot.getCanonicalPath());

            VirtualFile netbeansPropertiesVfLocal = LocalFileSystem.getInstance().findFileByPath(netbeansPropertiesPath);

            if (netbeansPropertiesVfLocal!= null) {
                try {
                    Properties properties = new Properties();
                    InputStream is = new ByteArrayInputStream(netbeansPropertiesVfLocal.contentsToByteArray());
                    properties.load(is);

                    Set<String> propertyNames = properties.stringPropertyNames();
                    for(String name: propertyNames) {
                        if(name.startsWith("file.reference") || name.equals("javac.classpath")) {
                            properties.remove(name);
                        }
                    }

                    return properties;
                } catch (IOException e) {
                    logger.error("Unable to read Netbeans property file. Stop process!");
                }
            } else {
                logger.error("Netbeans property file not found. Stop process!");
            }
        }
        return null;
    }
}



src/NetbeansModuleComponent.java
import com.intellij.ProjectTopics;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleComponent;
import com.intellij.openapi.module.ModuleUtilCore;
import com.intellij.openapi.project.ModuleListener;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileEvent;
import com.intellij.openapi.vfs.VirtualFileListener;
import com.intellij.packaging.impl.artifacts.ArtifactManagerImpl;
import com.intellij.packaging.impl.artifacts.ArtifactVirtualFileListener;
import com.intellij.util.xml.Convert;
import converter.ConverterFactory;
import converter.IntellijModuleImlChecksum;
import converter.ProjectFileConverter;
import org.jetbrains.annotations.NotNull;
import util.NotificationUtil;

import java.io.File;
import java.io.IOException;

/**
 * Created by trangdp on 17/05/2017.
 */
public class NetbeansModuleComponent implements ModuleComponent {
    private final Logger logger = Logger.getInstance("com.trangdp.plugin.netbeans.intellij");

    private Module myModule;

    public NetbeansModuleComponent(Module module) {
        MyModuleListener myModuleListener = new MyModuleListener();
        module.getProject().getMessageBus().connect().subscribe(ProjectTopics.MODULES, myModuleListener);
        VirtualFileListener f = new MyArtifactVirtualFileListener(module.getProject(), new ArtifactManagerImpl(module.getProject()));
        LocalFileSystem.getInstance().addVirtualFileListener(f);

        this.myModule = module;
    }

    @Override
    public void initComponent() {
        // TODO: insert component initialization logic here
    }

    @Override
    public void disposeComponent() {
        // TODO: insert component disposal logic here
    }

    @Override
    @NotNull
    public String getComponentName() {
        return "NetbeansModuleComponent";
    }

    @Override
    public void moduleAdded() {
        // Invoked when the module corresponding to this component instance has been completely
        // loaded and added to the project.
        NotificationUtil.notify("module component >>> event triggered");

        Project project = myModule.getProject();

        if (myModule.getModuleFile() != null) {
            ApplicationManager.getApplication().invokeLater(() -> {

                try {
                    LocalFileSystem.getInstance().addVirtualFileListener(new MyArtifactVirtualFileListener(project, new ArtifactManagerImpl(project)));

                    String relPath = myModule.getModuleFilePath().substring(0, myModule.getModuleFilePath().lastIndexOf(File.separator));
                    VirtualFile moduleRoot = project.getBaseDir().getFileSystem().findFileByPath(relPath);
                    if (moduleRoot.isDirectory()) {
                        String netbeansPropertiesPath = String.format("%s/nbproject/project.properties", moduleRoot.getCanonicalPath());

                        VirtualFile netbeansPropertiesVfLocal = LocalFileSystem.getInstance().findFileByPath(netbeansPropertiesPath);

                        if(netbeansPropertiesVfLocal != null) {
                            String content = new String(netbeansPropertiesVfLocal.contentsToByteArray());
                            ProjectFileConverter netbeans = new ConverterFactory().getConverter("netbeans-to-intellij");
                            netbeans.setModule(myModule);
                            netbeans.setProjectFileContent(content);
                            netbeans.convert();
                        }

                    }
                } catch (Exception e) {
                    logger.error(e.getMessage(), e);
                }
            });
        }
    }

    private class MyModuleListener implements ModuleListener {
        @Override
        public void moduleRemoved(@NotNull Project project, @NotNull Module module) {
            myModule = null;

            NotificationUtil.notify("Module " + module.getName() + " removed");
        }

        @Override
        public void moduleAdded(@NotNull Project project, @NotNull Module module) {
            myModule = module;

        }
    }

    private class MyArtifactVirtualFileListener extends ArtifactVirtualFileListener {
        @NotNull
        Project project;

        public MyArtifactVirtualFileListener(Project project, ArtifactManagerImpl artifactManager) {
            super(project, artifactManager);

            this.project = project;
        }

        @Override
        public void fileCreated(@NotNull VirtualFileEvent event) {
            super.fileCreated(event);
            logger.debug(event.getFileName());

            Module module = ModuleUtilCore.findModuleForFile(event.getFile(), project);

            if(module != null && event.getFileName().equals(module.getName() + ".iml")) {
                NotificationUtil.notify("file created=" + event.getFileName());

                ApplicationManager.getApplication().invokeLater(() -> {
                    Project project = module.getProject();

                    if(project.isDisposed()) {
                        logger.warn("Project " + project.getName() + " is disposed");
                        return;
                    }

                    try {
                        String relPath = module.getModuleFilePath().substring(0, module.getModuleFilePath().lastIndexOf(File.separator));
                        VirtualFile moduleRoot = project.getBaseDir().getFileSystem().findFileByPath(relPath);
                        if (moduleRoot.isDirectory()) {
                            String netbeansPropertiesPath = String.format("%s/nbproject/project.properties", moduleRoot.getCanonicalPath());

                            VirtualFile netbeansPropertiesVfLocal = LocalFileSystem.getInstance().findFileByPath(netbeansPropertiesPath);

                            if(netbeansPropertiesVfLocal != null) {
                                String content = new String(netbeansPropertiesVfLocal.contentsToByteArray());
                                ProjectFileConverter netbeans = new ConverterFactory().getConverter("netbeans-to-intellij");
                                netbeans.setModule(module);
                                netbeans.setProjectFileContent(content);
                                netbeans.convert();
                            }
                        }
                    } catch (Exception e) {
                        logger.error(e.getMessage(), e);
                    }
                });
            }
        }

        @Override
        public void contentsChanged(@NotNull VirtualFileEvent event) {
            super.contentsChanged(event);

            Module module = ModuleUtilCore.findModuleForFile(event.getFile(), project);

            //monitor content change for file <modulename>.iml
            if(module != null && event.getFileName().equals(module.getName() + ".iml")) {
                NotificationUtil.notify("file " + event.getFileName() + " content has changed");
                logger.info("Last updateLastChecksum file " + event.getFileName() + ":" + IntellijModuleImlChecksum.getInstance().getLastCheckSum(module.getName()));

                try {
                    String lastChecksum = IntellijModuleImlChecksum.getInstance().getLastCheckSum(module.getName());
                    String currentChecksum = IntellijModuleImlChecksum.getInstance().checksum(module.getModuleFilePath());
                    if(lastChecksum != null && !lastChecksum.isEmpty() && !currentChecksum.isEmpty() && !currentChecksum.equals(lastChecksum)) {
                        ProjectFileConverter converter = new ConverterFactory().getConverter("intellij-to-netbeans");
                        converter.setModule(module);
                        converter.setProjectFileContent(new String(module.getModuleFile().contentsToByteArray()));
                        converter.convert();
                    }

                    IntellijModuleImlChecksum.getInstance().updateLastChecksum(module.getName(), module.getModuleFilePath());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            //TODO: monitor content change for file nbproject/project.properties
        }
    }
}
